### PR TITLE
Unify rpc

### PR DESCRIPTION
**Pull request by marioiordanov**
## Usage related changes

Users can use http api endpoints via JSON-RPC api.

## Development related changes

- Endpoints accessible via HTTP endpoints can be invoked via JSON-RPC
- StarknetRequest -> JsonRpcRequest
- StarknetResponse -> JsonRpcResponse (StarknetResponse, DevnetResponse, Empty)
- Most of the http api methods logic is extracted into `<http api endpoint>_impl` function which is called from JSON-RPC counterpart
- Updated docs

## Checklist:

- [ ] Checked out the [contribution guidelines](CONTRIBUTING.md)
- [ ] Applied formatting - `./scripts/format.sh`
- [ ] No linter errors - `./scripts/clippy_check.sh`
- [ ] No unused dependencies - `./scripts/check_unused_deps.sh`
- [ ] No spelling errors - `./scripts/check_spelling.sh`
- [ ] Performed code self-review
- [ ] Rebased to the latest commit of the target branch (or merged it into my branch)
- [ ] Updated the docs if needed - `./website/README.md`
- [ ] Linked the [issues](https://github.com/0xSpaceShard/starknet-devnet-rs/issues) resolvable by this PR - [linking info](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- [ ] Updated the tests if needed; all passing - [execution info](https://github.com/0xSpaceShard/starknet-devnet-rs/?tab=readme-ov-file#test-execution)

### CODE REVIEW

File: crates/starknet-devnet-server/src/api/json_rpc/mod.rs [hunk 1]
```diff
@@ -312,67 +345,141 @@ pub enum StarknetRequest {
...
     StopAutoImpersonate,
+
+    #[serde(rename = "devnet_isAlive", with = "empty_params")]
+    IsAlive,
+    #[serde(rename = "devnet_dump")]
+    Dump(DumpPath),
+    #[serde(rename = "devnet_load")]
+    Load(LoadPath),
+
+    #[serde(rename = "devnet_postmanLoad")]
```
Comment (FabijanC) [inline crates/starknet-devnet-server/src/api/json_rpc/mod.rs:356]: Why are these three methods separated with empty lines?

**Review by FabijanC — CHANGES_REQUESTED** [review]
1. Should we eventually document all JSON-RPC methods and parameters in a specification .json file, like Starknet does? We would then not include JSON-RPC  examples in these .md documentation files, but would just link to the specification .json.

2. What about testing? My idea is to convert all tests that relied on HTTP methods to rely on JSON-RPC methods. That would be in line with deprecating these HTTP methods and dropping their support in 2-3 months. For integration tests, it would include changing the utility methods to change the sent requests.

File: crates/starknet-devnet/tests/common/background_devnet.rs [hunk 2]
```diff
@@ -167,15 +168,27 @@ impl BackgroundDevnet {
...
+            .post_json_async(RPC_PATH, body_json)
+            .await
+            .map_err(|_| RpcError::internal_error())?;
+
+        if let Some(result) = json_rpc_result.get("result") {
+            Ok(result.clone())
+        } else if let Some(error) = json_rpc_result.get("error") {
+            Err(serde_json::from_value(error.clone()).unwrap())
+        } else {
+            Err(RpcError::internal_error())
```
Comment (FabijanC) [inline crates/starknet-devnet/tests/common/background_devnet.rs:190]: Isn't this hiding the real response we are getting? Is there an `RpcError` variant which can wrap the response, at least in the stringified form?

Comment (marioiordanov) [reply crates/starknet-devnet/tests/common/background_devnet.rs:190]: the real response is in the "error" entry of the json response. The internal_error indicates that this json response is malformed

Comment (FabijanC) [reply crates/starknet-devnet/tests/common/background_devnet.rs:190]: I am aware of the case when the real error is in a dedicated property. But this way we'll have the same behavior if the server responds with 500, 400 or returns an invalid response. It is testing code, so we won't be affecting users, it just might be harder for us to debug our tests in the future. And maybe these "hidden error" scenarios are not even possible, I don't know.

Comment (marioiordanov) [reply crates/starknet-devnet/tests/common/background_devnet.rs]: [46fb94a](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/500/commits/46fb94aada82088d4c08ca6c1105ac9b5f22788b)
[a7193a0](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/500/commits/a7193a08df3233f4ea15a5158d071981bb79b052)

File: crates/starknet-devnet-server/src/api/json_rpc/mod.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet-devnet-server/src/api/json_rpc/mod.rs]: [625f9ae](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/500/commits/625f9aefaf26253b7f9e1642f9f817ee96ad1f81)

File: website/docs/balance.md [hunk 3]
```diff
@@ -39,3 +53,17 @@ Check the balance of an address by sending a `GET` request to `/account_balance`
 ```
 GET /account_balance?address=<ADDRESS>[&unit=<FRI|WEI>][&block_tag=<latest|pending>]
 ```
+
+```
+JSON-RPC
+{
+    "jsonrpc": "2.0",
+    "id": "1",
+    "method": "devnet_accountBalance",
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply website/docs/balance.md]: [639a08b](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/500/commits/639a08b5224be43e44e5db65a586b61e1ee8f3fe)

File: website/docs/server-config.md [hunk 4]
```diff
@@ -61,8 +61,17 @@ $ starknet-devnet --request-body-size-limit <BYTES>
...
 ```
 $ curl localhost:5050/config | jq .server_config
 ```
+
+```
+JSON-RPC
+{
+    "jsonrpc": "2.0",
+    "id": "1",
+    "method": "devnet_config"
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply website/docs/server-config.md]: [639a08b](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/500/commits/639a08b5224be43e44e5db65a586b61e1ee8f3fe)

File: website/docs/predeployed.md [hunk 5]
```diff
@@ -24,4 +24,4 @@ If you want to deploy an instance of an account contract class not predeclared o
 
 ## How to get predeployment info?
 
-The predeployment information is logged on Devnet startup. Predeployed accounts can be retrieved in JSON format by sending a `GET` request to `/predeployed_accounts` of your Devnet.
+The predeployment information is logged on Devnet startup. Predeployed accounts can be retrieved in JSON format by sending a `GET` request to `/predeployed_accounts` of your Devnet or `JSON-RPC` request with method name `devnet_predeployedAccounts`.
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply website/docs/predeployed.md]: [639a08b](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/500/commits/639a08b5224be43e44e5db65a586b61e1ee8f3fe)

**Review by FabijanC — APPROVED** [review]
Regarding the [question from my first review](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/500#pullrequestreview-2127227882):
> Should we eventually document all JSON-RPC methods and parameters in a specification .json file, like Starknet does?

I guess for now it stays as it is.

Thanks for addressing the change requests. Approving.

**Comment by marioiordanov** [comment]
For new JSON-RPC there is an issue created: https://github.com/0xSpaceShard/starknet-devnet-rs/issues/508

**Pull request merged**